File: C#/42.cs
public class Solution {
    public int Trap(int[] height) {
        int n = height.Length;
        if(n == 0) return 0;
        int left = 0, right = n-1;
        int leftMax = height[left], rightMax = height[right];
        int ans = 0;
        while(left < right){
            if(height[left] < height[right]){
                if(height[left] > leftMax) leftMax = height[left];
                else ans += leftMax - height[left];
                left++;
            }else{
                if(height[right] > rightMax) rightMax = height[right];
                else ans += rightMax - height[right];
                right--;
            }
        }
        return ans;
    }
}

File: python/42.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        
        n = len(height)
        left, right = [0]*n, [0]*n
        left[0], right[n-1] = height[0], height[n-1]
        
        for i in range(1, n):
            left[i] = max(left[i-1], height[i])
            
        for i in range(n-2, -1, -1):
            right[i] = max(right[i+1], height[i])
            
        ans = 0
        for i in range(n):
            ans += min(left[i], right[i]) - height[i]
            
        return ans